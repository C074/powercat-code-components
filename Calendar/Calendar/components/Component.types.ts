export interface ICalendarProps {
    width?: number;
    height?: number;
    themeJSON?: string;
    setFocus?: string;
    tabIndex?: number;
    selectedDateValue: Date | undefined;
    minDate: Date | undefined;
    maxDate: Date | undefined;
    ariaLabel: string;
    showGoToToday?: boolean;
    monthPickerVisible?: boolean;
    dayPickerVisible?: boolean;
    highlightSelectedMonth?: boolean;
    highlightCurrentMonth?: boolean;
    showSixWeeksByDefault?: boolean;
    showWeekNumbers: boolean;
    firstDayOfWeek: string;
    isDisabled: boolean;
<<<<<<< HEAD
    language: string;
=======
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
    // eslint-disable-next-line
    onSelected: (selectedDateValue: Date) => void;
    backgroundColor?: string;
}
