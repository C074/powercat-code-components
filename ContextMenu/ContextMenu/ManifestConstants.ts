export const enum ManifestPropertyNames {
    dataset = 'dataset',
}

export const enum ItemColumns {
    DisplayName = 'ItemDisplayName',
    Key = 'ItemKey',
    Enabled = 'ItemEnabled',
    IconName = 'ItemIconName',
    IconColor = 'ItemIconColor',
    TextColor = 'ItemTextColor',
    BackgroundColor = 'ItemBackgroundColor',
    BorderColor = 'ItemBorderColor',
    ParentKey = 'ItemParentKey',
    IconOnly = 'ItemIconOnly',
    Visible = 'ItemVisible',
    Checked = 'ItemChecked',
}

export const enum InputEvents {
    SetFocus = 'SetFocus',
}

export const enum InputProperties {
    InputEvent = 'InputEvent',
}

export const TextAlignmentTypes = {
    '0': 'center',
    '1': 'left',
    '2': 'right',
};
