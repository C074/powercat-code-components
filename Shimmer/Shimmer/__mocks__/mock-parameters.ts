/* istanbul ignore file */

import { IInputs } from '../generated/ManifestTypes';
<<<<<<< HEAD
import { MockEnumProperty, MockStringProperty, MockWholeNumberProperty } from './mock-context';
=======
import { MockEnumProperty, MockStringProperty } from './mock-context';
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
import { MockDataSet } from './mock-datasets';

export function getMockParameters(): IInputs {
    return {
        Theme: new MockStringProperty(),
        AccessibilityLabel: new MockStringProperty(),
<<<<<<< HEAD
        RowCount: new MockWholeNumberProperty(),
        items: new MockDataSet([]),
=======
        items: new MockDataSet([]),
        rows: new MockDataSet([]),
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
        SpacebetweenShimmer: new MockEnumProperty(),
    };
}
