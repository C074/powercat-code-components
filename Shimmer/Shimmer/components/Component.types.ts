<<<<<<< HEAD
import { IShimmerElement } from '@fluentui/react';

=======
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
export interface ICustomShimmerItem {
    id: string;
    type: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    key: string;
    height: number;
    width: string;
    verticalAlign: string;
    rowkey: string;
}

export interface IShimmerRows {
    id: string;
    key: string;
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    data: any;
    order: number;
    count: number;
    width: number;
}

export interface IShimmerProps {
    width?: number;
    height?: number;
<<<<<<< HEAD
    shimmerElements: IShimmerElement[];
    themeJSON?: string;
    spacebetweenShimmer?: string;
    rowCount?: number;
=======
    items: ICustomShimmerItem[];
    rowDetails: IShimmerRows[];
    themeJSON?: string;
    spacebetweenShimmer?: string;
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
}
