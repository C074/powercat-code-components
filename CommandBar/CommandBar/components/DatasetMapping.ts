import { CanvasCommandItem } from './Component.types';
import {
    getColorFromString,
    getShade,
    IButtonStyles,
    ICommandBarItemProps,
    IContextualMenuItem,
    IIconProps,
    Shade,
} from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';

export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): CanvasCommandItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getFormattedValue(ItemColumns.DisplayName),
            enabled: defaultIfNull(record.getValue(ItemColumns.Enabled) as boolean, true),
            iconName: record.getFormattedValue(ItemColumns.IconName),
            iconColor: record.getFormattedValue(ItemColumns.IconColor),
            farItem: undefinedIfNull(record.getValue(ItemColumns.FarItem) as boolean),
            overflow: undefinedIfNull(record.getValue(ItemColumns.Overflow) as boolean),
            parentItemKey: record.getFormattedValue(ItemColumns.ParentKey),
            iconOnly: undefinedIfNull(record.getValue(ItemColumns.IconOnly) as boolean),
            checked: undefinedIfNull(record.getValue(ItemColumns.Checked) as boolean),
            visible: defaultIfNull(record.getValue(ItemColumns.Visible) as boolean, true),
            split: undefinedIfNull(record.getValue(ItemColumns.Split) as boolean),
            data: record,
        } as CanvasCommandItem;
    });
}

function getDummyAction(key: string): CanvasCommandItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as CanvasCommandItem;
}

export function getCommandsWithChildren(
    items: CanvasCommandItem[],
    disabled: boolean,
    onClick: (ev?: unknown, item?: IContextualMenuItem | undefined) => boolean,
): ICommandBarItemProps[] {
    return items
        .filter((i) => !i.parentItemKey && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
}

export function getCommandBarItemProps(
    items: CanvasCommandItem[],
    item: CanvasCommandItem,
    disabled: boolean,
    onClick: (ev?: unknown, item?: IContextualMenuItem | undefined) => boolean,
): ICommandBarItemProps {
    const subItems = items
        .filter((i) => i.parentItemKey === item.key && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });

    const buttonStyles = {
        root: {
            background: 'rgba(255, 255, 255,0)',
        },
    } as IButtonStyles;

    const iconProps =
        item.iconName &&
        ({
            iconName: item.iconName,
        } as IIconProps);

    // If custom icon color, find the shade slots for the hover/pressed state
    if (iconProps && item.iconColor) {
        const iconColor = item.iconColor;
        iconProps.styles = { root: { color: item.iconColor + ' !important' } };

        // Attempt to set a darker shade for hover/pressed colors
        const primaryColor = getColorFromString(iconColor);
        if (primaryColor) {
            const iconColorDarker = getShade(primaryColor, Shade.Shade8);
            const secondaryColorStyle = { color: iconColorDarker?.str + ' !important' };
            buttonStyles.iconHovered = secondaryColorStyle;
            buttonStyles.iconPressed = secondaryColorStyle;
            buttonStyles.iconExpanded = secondaryColorStyle;
            buttonStyles.iconChecked = secondaryColorStyle;
            buttonStyles.iconExpandedHovered = secondaryColorStyle;
        }
    }

    return {
        key: item.key,
        text: item.name,
        disabled: item.enabled === false || disabled,
        onClick: onClick,
        iconOnly: item.iconOnly === true,
        iconProps: iconProps,
        split: item.split,
        canCheck: item.checked !== undefined,
        checked: item.checked,
        data: item,
        buttonStyles: buttonStyles,
        subMenuProps:
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined,
    } as ICommandBarItemProps;
}

function undefinedIfNull<T>(value: T | null) {
    return defaultIfNull(value, undefined);
}
function defaultIfNull<T>(value: T | null, defaultValue: T) {
    return value === null ? defaultValue : value;
}
