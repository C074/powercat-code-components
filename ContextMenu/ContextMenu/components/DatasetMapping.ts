import { CanvasContextMenuItem } from './Component.types';
import { IContextualMenuItem } from '@fluentui/react';
import { ItemColumns } from '../ManifestConstants';

export function getMenuItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): CanvasContextMenuItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyAction('1'), getDummyAction('2'), getDummyAction('3')];
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            name: record.getValue(ItemColumns.DisplayName) as string,
            enabled: record.getValue(ItemColumns.Enabled) as boolean,
            iconName: record.getValue(ItemColumns.IconName) as string,
            iconColor: record.getValue(ItemColumns.IconColor) as string,
            textColor: record.getValue(ItemColumns.TextColor) as string,
            borderColor: record.getValue(ItemColumns.BorderColor) as string,
            backgroundColor: record.getValue(ItemColumns.BackgroundColor) as string,
            parentItemKey: record.getValue(ItemColumns.ParentKey) as string,
            iconOnly: record.getValue(ItemColumns.IconOnly) as boolean,
            checked: record.getValue(ItemColumns.Checked) as boolean,
            visible: record.getValue(ItemColumns.Visible) as boolean,
            data: record,
        } as CanvasContextMenuItem;
    });
}

function getDummyAction(key: string): CanvasContextMenuItem {
    return {
        id: key,
        key: key,
        name: 'Item ' + key,
        iconName: 'Unknown',
    } as CanvasContextMenuItem;
}

export function getCommandsWithChildren(
    items: CanvasContextMenuItem[],
    disabled: boolean,
    onClick: (ev?: unknown, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem[] {
    return items
        .filter((i) => !i.parentItemKey && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
}

export function getCommandBarItemProps(
    items: CanvasContextMenuItem[],
    item: CanvasContextMenuItem,
    disabled: boolean,
    onClick: (ev?: unknown, item?: IContextualMenuItem | undefined) => boolean,
): IContextualMenuItem {
    const subItems = items
        .filter((i) => i.parentItemKey === item.key && i.visible !== false)
        .map((i) => {
            return getCommandBarItemProps(items, i, disabled, onClick);
        });
    return {
        key: item.key,
        text: item.name,
        disabled: item.enabled === false || disabled,
        onClick: onClick,
        iconOnly: item.iconOnly === true,
        iconProps: {
            iconName: item.iconName,
            styles: { root: { color: item.iconColor } },
        },
        checked: item.checked,
        data: item,
        buttonStyles: { root: { background: 'rgba(255, 255, 255,0)' } },
        subMenuProps:
            subItems.length > 0
                ? {
                      items: subItems,
                  }
                : undefined,
    } as IContextualMenuItem;
}
