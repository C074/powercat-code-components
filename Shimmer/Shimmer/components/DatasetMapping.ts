import { ICustomShimmerItem, IShimmerRows } from './Component.types';
import { ItemColumns, RowColumns } from '../ManifestConstants';
import { IShimmerElement, ShimmerElementType } from '@fluentui/react';
export function getItemsFromDataset(dataset: ComponentFramework.PropertyTypes.DataSet): ICustomShimmerItem[] {
    if (dataset.error || dataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return getDummyItems(0).concat(getDummyItems(1));
    }
    const keyIndex: Record<string, number> = {};
    return dataset.sortedRecordIds.map((id) => {
        const record = dataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(ItemColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: key,
            type: record.getValue(ItemColumns.Type) as string,
            height: record.getValue(ItemColumns.Height) as number,
            width: record.getValue(ItemColumns.Width) as string,
            verticalAlign: record.getValue(ItemColumns.VerticalAlign) as string,
            rowkey: record.getValue(ItemColumns.RowKey) as string,
            data: record,
        } as ICustomShimmerItem;
    });
}

export function getRowDetailsFromDataset(rowdataset: ComponentFramework.PropertyTypes.DataSet): IShimmerRows[] {
    if (rowdataset.error || rowdataset.paging.totalResultCount === undefined) {
        // Dataset is not defined so return dummy items
        return [getDummyRow('1')];
    }
    const keyIndex: Record<string, number> = {};
    return rowdataset.sortedRecordIds.map((id) => {
        const record = rowdataset.records[id];
        // Prevent duplicate keys by appending the duplicate index
        let key = record.getValue(RowColumns.Key) as string;
        if (keyIndex[key] !== undefined) {
            keyIndex[key]++;
            key += `_${keyIndex[key]}`;
        } else keyIndex[key] = 1;
        return {
            id: record.getRecordId(),
            key: record.getValue(RowColumns.Key) as string,
            count: record.getValue(RowColumns.Count) as number,
            order: record.getValue(RowColumns.Order) as number,
            width: record.getValue(RowColumns.Width) as number,
            data: record,
        } as IShimmerRows;
    });
}

export function getShimmerElements(items: ICustomShimmerItem[], rowKey: string): IShimmerElement[] {
    return items
        .filter((item) => item.rowkey === rowKey)
        .map((item: ICustomShimmerItem) => ({
            type: getShimmerElementType(item.type),
            ...(typeof item.width !== 'undefined' && item.width !== null && { width: item.width + '%' }),
            ...(typeof item.height !== 'undefined' && item.height !== null && { height: item.height }),
            ...(typeof item.verticalAlign !== 'undefined' &&
                item.verticalAlign !== null && { width: item.verticalAlign }),
        })) as IShimmerElement[];
}

function getDummyRow(key: string): IShimmerRows {
    return {
        id: key,
        key: key,
        count: key === '1' ? 5 : 1,
        order: parseInt(key),
        width: 100,
    } as IShimmerRows;
}

export function getShimmerElementType(shimmerElementType: string): ShimmerElementType {
    let currentElementType: ShimmerElementType = ShimmerElementType.line;
    switch (shimmerElementType.toLowerCase()) {
        case 'circle':
            currentElementType = ShimmerElementType.circle;
            break;
        case 'gap':
            currentElementType = ShimmerElementType.gap;
            break;
        default:
            currentElementType;
    }
    return currentElementType;
}

function getDummyItems(index: number): ICustomShimmerItem[] {
    return [
        {
            id: 1,
            key: 1,
            width: index === 0 ? '3.8' : '1.8',
            height: index === 0 ? 60 : 20,
            rowkey: '1',
            type: index === 0 ? 'circle' : 'line',
        },
        {
            id: 2,
            key: 2,
            width: '2',
            height: 10,
            rowkey: '1',
            type: 'gap',
        },
        {
            id: 3,
            key: 3,
            width: '20',
            height: 10,
            rowkey: '1',
            type: 'line',
        },
        {
            id: 4,
            key: 4,
            width: '2',
            height: 10,
            rowkey: '1',
            type: 'gap',
        },
        {
            id: 5,
            key: 5,
            width: '2.8',
            height: 30,
            rowkey: '1',
            type: 'line',
        },
        {
            id: 6,
            key: 6,
            width: '2',
            height: 10,
            rowkey: '1',
            type: 'gap',
        },
        {
            id: 7,
            key: 7,
            width: '20',
            height: 10,
            rowkey: '1',
            type: 'line',
        },
        {
            id: 8,
            key: 8,
            width: '1',
            height: 10,
            rowkey: '1',
            type: 'gap',
        },
    ] as unknown as ICustomShimmerItem[];
}
