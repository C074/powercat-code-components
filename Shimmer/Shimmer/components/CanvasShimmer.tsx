import * as React from 'react';
import { Shimmer, ThemeProvider, createTheme, IPartialTheme, mergeStyles } from '@fluentui/react';
import { IShimmerProps } from './Component.types';
<<<<<<< HEAD

export const CanvasShimmer = React.memo((props: IShimmerProps) => {
    const { shimmerElements, themeJSON, spacebetweenShimmer, rowCount } = props;
=======
import { getShimmerElements } from './DatasetMapping';

export const CanvasShimmer = React.memo((props: IShimmerProps) => {
    const { items, themeJSON, spacebetweenShimmer, rowDetails } = props;
    const shimmerElements: JSX.Element[] = [];
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
    const wrapperClass = mergeStyles({
        padding: 2,
        selectors: {
            '& > .ms-Shimmer-container': {
                margin: spacebetweenShimmer?.concat(' 0'),
            },
        },
<<<<<<< HEAD
        width:props.width
=======
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
    });

    const theme = React.useMemo(() => {
        try {
            return themeJSON ? createTheme(JSON.parse(themeJSON) as IPartialTheme) : undefined;
        } catch (ex) {
            /* istanbul ignore next */
            console.error('Cannot parse theme', ex);
        }
    }, [themeJSON]);

<<<<<<< HEAD
    return (
        <ThemeProvider theme={theme} className={wrapperClass}>
            {[...Array(rowCount)].map((key) => {
                return <Shimmer key={key} shimmerElements={shimmerElements} />;
            })}
=======
    rowDetails
        .sort((prevRow, currentRow) => prevRow.order - currentRow.order)
        .map((row) => {
            for (let i = 0; i < row.count; i++) {
                shimmerElements.push(
                    <Shimmer
                        width={row.width ? row.width.toString().concat('%') : '100%'}
                        shimmerElements={getShimmerElements(items, row.key)}
                    />,
                );
            }
        });

    return (
        <ThemeProvider theme={theme} className={wrapperClass}>
            {shimmerElements.length > 0 && shimmerElements}
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
        </ThemeProvider>
    );
});
CanvasShimmer.displayName = 'CanvasShimmer';
