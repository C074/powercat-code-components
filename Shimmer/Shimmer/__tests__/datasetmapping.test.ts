<<<<<<< HEAD
import { getItemsFromDataset, getShimmerElements, getShimmerElementType } from '../components/DatasetMapping';
import { ItemColumns } from '../ManifestConstants';
=======
import {
    getRowDetailsFromDataset,
    getItemsFromDataset,
    getShimmerElements,
    getShimmerElementType,
} from '../components/DatasetMapping';
import { ItemColumns, RowColumns } from '../ManifestConstants';
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
import { MockDataSet, MockEntityRecord } from '../__mocks__/mock-datasets';
import { ShimmerElementType } from '@fluentui/react';

describe('DatasetMapping', () => {
    beforeEach(() => jest.clearAllMocks());

    it('returns correct props', () => {
        const items = [
            new MockEntityRecord('1', {
                [ItemColumns.Key]: '1',
                [ItemColumns.Height]: 10,
                [ItemColumns.Width]: '20',
                [ItemColumns.RowKey]: '1',
                [ItemColumns.Type]: 'gap',
            }),
        ];
<<<<<<< HEAD
        expect(getShimmerElementType('Circle')).toEqual(ShimmerElementType.circle);
        const shimmerItems = getShimmerElements(getItemsFromDataset(new MockDataSet(items)));
        expect(shimmerItems).toMatchSnapshot();
=======

        const rowItems = [
            new MockEntityRecord('1', {
                [RowColumns.Key]: '1',
                [RowColumns.Order]: 1,
                [RowColumns.Count]: 5,
                [RowColumns.Width]: 100,
            }),
        ];

        expect(getShimmerElementType('Circle')).toEqual(ShimmerElementType.circle);
        const rowDetails = getRowDetailsFromDataset(new MockDataSet(rowItems));
        expect(rowDetails).toHaveLength(1);
        const shimmerItems = getShimmerElements(getItemsFromDataset(new MockDataSet(items)), rowDetails[0].key);
        expect(shimmerItems).toMatchSnapshot();
        expect(rowDetails).toMatchSnapshot();
>>>>>>> 48d69ffc393a155d03b3e7517eb7dc7512b10fd5
    });
});
