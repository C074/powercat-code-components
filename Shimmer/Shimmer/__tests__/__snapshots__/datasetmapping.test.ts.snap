// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DatasetMapping returns correct props 1`] = `
Array [
  Object {
    "height": 10,
    "type": 3,
    "width": "20%",
  },
]
`;

exports[`DatasetMapping returns correct props 2`] = `
Array [
  Object {
    "count": 5,
    "data": MockEntityRecord {
      "getNamedReference": [MockFunction],
      "id": "1",
      "values": Object {
        "RowCount": 5,
        "RowKey": "1",
        "RowOrder": 1,
        "RowWidth": 100,
      },
    },
    "id": "1",
    "key": "1",
    "order": 1,
    "width": 100,
  },
]
`;
